pycloudlib/ibm/cloud.py

```diff
@@ -381,16 +333,28 @@ def launch(
             **kwargs,
         )
 
-        instance = IBMInstance.with_floating_ip(
+        instance: IBMInstance = IBMInstance.from_raw_instance(
             self.key_pair,
             client=self._client,
             instance=raw_instance,
-            floating_ip=floating_ip,
             username=username,
             using_existing_floating_ip=floating_ip_substring is not None,
         )
+
+        # append before attaching floating ip so we can clean up in case of failure while attaching floating ip
         self.created_instances.append(instance)
 
+        # floating ip stuff
+        if floating_ip_substring:
+            instance.attach_floating_ip_until_success(
```

Comment:
What if we have only one free ip but two instances get to this point? One of them will never success here, right?

What about something like the following?

```python
        need_floating_ip = not floating_ip_substring
        if floating_ip_substring:
            try:
                instance.attach_floating_ip_until_success(
                    floating_ip_name_includes=floating_ip_substring,
                    zone=self.zone,
                )
            except Exception as e:
                need_floating_ip = True
        if need_floating_ip:
            self._log.info("Creating new floating ip.")
            floating_ip = self._create_floating_ip(name=floating_ip_name)
            instance.attach_floating_ip(floating_ip=floating_ip)

```

Reply:
This is handled in the downstream function _choose_from_existing_floating_ips(). This function will raise a capacity error once there are no more available floating IPs. This is important because it allows for consumers of pycloudlib to do retries (like CTF). Doing it this way is also important because if the user specifies that an existing floating IP should be used, we do NOT want to spin up a new one for them since that would be unwanted behavior.